### ISSUE TITLE
Using convolution layers with VAE

### ISSUE THREAD

**Issue by pranshu97** [probtorch/probtorch#50]
I am trying to use the semi supervised approach to train a Unet like model(without the cross connections) with latent space as the bridge. But the reconstruction I am getting is very poor (sometimes just noise), also I am facing exploding gradient problem often.
There are no example showing how I can use convolutions with the VAE, is there a reason for that? Or am I missing something?

**Comment by babak0032** [comment]
Not sure if I can pinpoint the problem without seeing the code first. Implementation-wise, there is almost no difference between using cnns and mlps (other than the input shape). Maybe the one thing that you should be careful in probtorch when using cnns is the @expand_inputs function. A cnn expects a 4-dimensional tensor is if you're expanding your input from (B,C,H,W) to (sample_size, B,C,H,W), then make sure you reshape it to (sample_size*B,C,H,W) before feeding it to the network and then reshape it back at the end. But you won't run into this problem if your sample_size is 1. 

Some of the other things to watch out for if you're having poor reconstructions are:
- Make sure you constrain the domain of your recons by using a Sigmoid layer at the end of the decoder. 
- If training is unstable, you can try adding a bit of noise to the variance in latent space (e.g. styles_std = torch.exp(self.style_log_std(hiddens2)) + 1e-5)
- Try increasing the size of your latent space. 
- Try tuning down your beta. 

But again, I can't say more without seeing the code first. 

One more note: I haven't trained a Unet like model myself but I think the cross connection might be necessary to get a good reconstructions given the huge size of the input, though I'm not sure.

**Comment by pranshu97** [comment]
I am using a sample size of 1. Will try tuning beta and latent space. Thanks for you help.
